File: kamacoder108/kamacoder108.py
father = list()

def find(u):
    if u == father[u]:
        return u
    else:
        father[u] = find(father[u])
        return father[u]
        
def is_same(u, v):
    u = find(u)
    v = find(v)
    return u == v
    
def join(u, v):
    u = find(u)
    v = find(v)
    if u != v:
        father[u] = v
        
if __name__ == "__main__":

    n = int(input())
    for i in range(n + 1):
        father.append(i)
    
    result = None
    for i in range(n):
        s, t = map(int, input().split())
        if is_same(s, t):
            result = str(s) + ' ' + str(t)
        else:
            join(s, t)
        
    
    print(result)

File: kamacoder108/kamacoder108.cpp
#include<iostream>
#include<vector>
using namespace std;

int n;
vector<int> father(1001, 0);

void init(){
    for (int i = 0; i <= n; i++) father[i] = i;
}

int find(int u){
    return u == father[u] ? u : father[u] = find(father[u]);
}

bool isSame(int u, int v){
    u = find(u);
    v = find(v);
    return u == v;
}

void join(int u, int v){
    u = find(u);
    v = find(v);
    if (u == v) return;
    father[v] = u;
}

int main(){
    int s, t;
    cin >> n;
    init();
    for (int i = 0; i < n; i++){
        cin >> s >> t;
        if (isSame(s, t)) {
            cout << s << " " << t << endl;
        } else{
            join(s, t);
        }
    }
}